File: src/SolverHeuristic.cpp
#include "../include/SolverHeuristic.h"


using std::vector;


// constructors and destructors
SolverHeuristic::SolverHeuristic(// misc parameters
                   bool silent,

                   // GA settings
                   int population_size,
                   int num_generations,
                   int improvement_depth,
                   int elite_size,
                   double distance_threshold,

                   double crossover_prob,
                   double mutation_prob,
                   double improvement_prob,
                   unsigned long seeds[6]) : Solver(silent)
{
    assert(distance_threshold >= 0);
    population_size_ = population_size;
    elite_size_ = elite_size;
    distance_threshold_ = distance_threshold;
    num_generations_ = num_generations;
    improvement_depth_ = improvement_depth;

    crossover_prob_ = crossover_prob;
    mutation_prob_ = mutation_prob;
    improvement_prob_ = improvement_prob;
    seeds_ = seeds;

}

SolverHeuristic::~SolverHeuristic() { }


Solution uniform_crossover(Solution parent1,
                           Solution parent2,
                           double crossover_probability,
                           RngStream* prng)
{
    // uniform crossover operator over the genes of two solutions
    vector<int> inherited_genes = vector<int>();
    double prob; // Uniform(0, 1)
    for (size_t index = 0; index != parent1.get_genes().size(); ++index)
    {
        prob = prng->RandU01();
        if (prob < crossover_probability)
        {
            inherited_genes.push_back(parent1.get_genes()[index]);
        }
        else
        {
            inherited_genes.push_back(parent2.get_genes()[index]);
        }
    }
    return Solution(inherited_genes);
}


Solution uniform_mutate(Solution& sol,
                         std::vector<Range> ranges,
                         RngStream *prng,
                         double mutation_probability=0.1)
{
    // mutate the solution by simply swapping with a probability
    double prob;
    vector<int> mutated_genes = vector<int>(sol.get_genes().size(), 0);

    for (size_t index = 0; index != sol.get_genes().size(); ++index)
    {
        prob = prng->RandU01();
        if (prob < mutation_probability)
        {
            // exchange for a random gene within the same range
            mutated_genes[index] = prng->RandInt(ranges[index].low,
                                                 ranges[index].high-1);
        }
        else
        {
            // keep the same gene
            mutated_genes[index] = sol.get_genes()[index];
        }
    }
    return Solution(mutated_genes);
}


Range select_2(int low, int high, RngStream *prng)
{
    // select 2 different random integers in the interval
    assert(high - 1 > low);
    int first = prng->RandInt(low, high-1);
    int second = prng->RandInt(low, high-1);

    while (first == second)
    {
        second = prng->RandInt(low, high-1);
    }
    Range ret = {first, second};
    return ret;
}


vector<Solution> binary_tournament_selection(vector<Solution> population,
                                             int num_to_select,
                                             RngStream *prng)
{
    // classical binary tournament selection
    assert(num_to_select > 0);
    vector<Solution> selected = vector<Solution>();
    Range indices;
    int pop_size = population.size() - 1;
    for (int index = 0; index != num_to_select; ++index)
    {
        indices = select_2(0, pop_size, prng);
        if (population[indices.low].get_score() < population[indices.high].get_score())
        {
            selected.push_back(Solution(population[indices.low]));
        }
        else
        {
            selected.push_back(Solution(population[indices.high]));
        }
    }
    return selected;
}



std::vector<Solution> SolverHeuristic::solve(std::vector< std::vector<double> > distance_matrix,
                                             std::vector<Range> ranges) const
{
    // solve the consensus problem using a genetic algorithm

    // adding an assert message by using the trick from
    //  http://stackoverflow.com/questions/3692954/add-custom-messages-in-assert
    assert(population_size_ > 0 && "Illegal population size");
    assert(num_generations_ > 0 && "Illegal number of generations");
    assert(elite_size_ >= 0 && "Illegal elite size (below zero)");
    assert(elite_size_ <= population_size_ && "Illegal elite size (bigger than population");
    assert(0 <= crossover_prob_ && "Illegal crossover probability (lower than zero)");
    assert(crossover_prob_ <= 1 && "Illegal crossover probability (higher than one)");
    assert(0 <= mutation_prob_ && "Illegal mutation probability (lower than zero)");
    assert(mutation_prob_ <= 1 && "Illegal mutation probability (higher than one)");

    // seed the pseudorandom generator (MRG32k3a from L'Ecuyer)
    RngStream * prng = new RngStream();
    prng->SetSeed(seeds_);

    // some declarations for later
    vector< vector<Solution> > best_solutions = vector< vector<Solution> >();
    vector<Solution> current_best_solutions;
    vector<Solution> children, parents, elite;
    Solution parent1;
    Solution parent2;
    double current_best_score;
    double scaled_threshold = distance_threshold_ * ranges.size() * (ranges.size()-1);
    vector<Solution>::iterator sol_it;

    // start the progress meter
    ProgressBar bar = ProgressBar();

    // initialize the population
    vector<Solution> population = initialize_population(ranges, population_size_, prng);

    // main loop
    for (size_t generation_index = 0; generation_index != num_generations_; ++generation_index)
    {

        // output the progress bar if not silent
        if (! silent_)
            bar.update( ((float) generation_index) / num_generations_);

        // score the solutions and sort the population by score
        for (sol_it = population.begin(); sol_it != population.end(); ++sol_it)
        {
            assign_pairwise_distance_score(*sol_it, distance_matrix);
        }
        std::sort(population.begin(), population.end());

        // remember the best solutions of the current generation
        current_best_solutions = vector<Solution>();
        current_best_score = population[0].get_score(); // because it is sorted
        for (sol_it = population.begin(); sol_it != population.end(); ++sol_it)
        {
            if (sol_it->get_score() <= current_best_score + scaled_threshold)
            {
                // try to find if it exists already in the elite vector
                bool found = false;
                for (size_t j = 0; j != current_best_solutions.size(); ++j)
                {
                    if (current_best_solutions[j] == *sol_it)
                    {
                        found = true;
                        break;
                    }
                }
                if (! found)
                {
                    current_best_solutions.push_back(Solution(*sol_it));
                }
            }
            else
            {
                break;
            }
        }

        best_solutions.push_back(current_best_solutions);

        // elitist selection with only unique individuals, no repetition
        elite = vector<Solution>();

        for (size_t i = 0; i < population.size(); ++i)
        {
            if (elite.size() == elite_size_)
            {
                break;
            }
            else
            {
                bool found = false;
                for(size_t j = 0; j < elite.size(); ++j)
                {
                    if ( (population[i] == elite[j]) )
                    {
                        found = true;
                        break;
                    }
                }
                if (! found)
                {
                    elite.push_back(Solution(population[i]));
                }
            }
        }

        // selection process
        parents = binary_tournament_selection(population, ((population_size_ - elite.size())*2), prng);
        children = vector<Solution>();
        for (size_t i = 0; i != population_size_ - elite_size_; ++i)
        {
            parent1 = parents[i*2];
            parent2 = parents[(i*2)+1];
            Solution child;

            // crossover
            if (prng->RandU01() < crossover_prob_)
            {
                child = uniform_crossover(parent1, parent2, 0.5, prng);
            }
            else
            {
                child = Solution(parent1);
            }

            // mutation
            if (prng->RandU01() < mutation_prob_)
                uniform_mutate(child, ranges, prng);

            // improvement
            if (prng->RandU01() < improvement_prob_)
                steepest_improvement(child, distance_matrix, ranges, improvement_depth_);

            // children is done
            children.push_back(child);
        }

        // replace the population by its children and the previous elite
        for (sol_it = elite.begin(); sol_it != elite.end(); ++sol_it)
        {
            children.push_back(Solution(*sol_it));
        }

        // swap the two populations
        population = children;
    }

    // clean the progress bar
    if (! silent_)
        bar.clean();

    // cleanup
    delete prng;

    // keep all the unique best solutions up to a specified suboptimal threshold
    std::vector<Solution> suitable_solutions = std::vector<Solution>();
    double score_threshold = current_best_score + scaled_threshold;
    for (size_t i = 0; i != best_solutions.size(); ++i)
    {
        for(size_t j = 0; j != best_solutions[i].size(); ++j)
        {
            if (best_solutions[i][j].get_score() <= score_threshold)
            {
                // verify if the solution is already in the optimal ones
                bool found = false;
                for(size_t k = 0; k < suitable_solutions.size(); ++k)
                {
                    if ( (best_solutions[i][j] == suitable_solutions[k]) )
                    {
                        found = true;
                        break;
                    }
                }
                if (! found)
                {
                    suitable_solutions.push_back(Solution(best_solutions[i][j]));
                }
            }
        }
    }
    return suitable_solutions;
}

File: src/src/mccons/solvers/SolverHeuristic.java
package mccons.solvers;

import java.util.ArrayList;
import java.util.Collections;

import mccons.util.Pair;
import mccons.util.ProgressBar;
import mccons.util.RngStream;


public class SolverHeuristic extends Solver {
    // misc parameters
    private boolean verbose;
    private double tolerance;
    private long[] seeds;

    // GA settings
    private int populationSize;
    private int numGenerations;
    private int eliteSize;

    private double improvementProbability;
    private int improvementDepth;

    private double crossoverProbability;
    private double crossoverMixingRatio;

    private double mutationProbability;
    private double mutationStrength;





    public SolverHeuristic(// misc parameters
                           boolean verbose_,
                           double tolerance_,
                           long[] seeds_,

                           // GA settings
                           int populationSize_,
                           int numGenerations_,
                           int eliteSize_,
                           double improvementProbability_,
                           int improvementDepth_,
                           double crossoverProbability_,
                           double crossoverMixingRatio_,
                           double mutationProbability_,
                           double mutationStrength_) {
        assert (tolerance_ >= 0);
        assert (seeds.length == 6);

        verbose = verbose_;
        tolerance = tolerance_;
        seeds = seeds_;
        populationSize = populationSize_;
        numGenerations = numGenerations_;
        eliteSize = eliteSize_;
        improvementProbability = improvementProbability_;
        improvementDepth = improvementDepth_;
        crossoverProbability = crossoverProbability_;
        crossoverMixingRatio = crossoverMixingRatio_;
        mutationProbability = mutationProbability_;
        mutationStrength = mutationStrength_;

    }

    @Override
    public boolean isVerbose() {
        return verbose;
    }

    /**
     * uniform crossover operator over the genes of two solutions
     *
     * @param parent1 first parent solution
     * @param parent2 second parent solution
     * @param stream pseudo-random generator stream
     * @return a crossover between both parents :P
     */
    public static Solution uniformCrossover(Solution parent1, Solution parent2, double mixingRatio, RngStream stream) {

        ArrayList<Integer> newGenes = parent1.getGenes();
        int size = parent1.getGenes().size();

        for (int index = 0; index != size; ++index) {
            if (stream.randU01() < mixingRatio) {
                newGenes.set(index, parent2.getGenes().get(index));
            }
        }
        return new Solution(newGenes, Double.POSITIVE_INFINITY);
    }


    /**
     * @param sol
     * @param ranges
     * @param mutation_probability
     * @param stream
     * @return
     */
    public Solution uniformMutate(Solution sol,
                                  ArrayList<Pair<Integer, Integer>> ranges,
                                  double mutation_probability,
                                  RngStream stream) {
        // mutate the solution by simply swapping with a probability
        ArrayList<Integer> mutated_genes = sol.getGenes();
        int gene_size = mutated_genes.size();
        for (int index = 0; index != gene_size; ++index) {
            if (stream.randU01() < mutation_probability) {
                // exchange for a random gene within the same range
                mutated_genes.set(index, stream.randInt(ranges.get(index).getFirst(), ranges.get(index).getSecond() - 1));
            }
        }
        return new Solution(mutated_genes, Double.POSITIVE_INFINITY);
    }


    Pair<Integer, Integer> select2(int low, int high, RngStream stream) {
        // select 2 different random integers in the interval
        assert (high - 1 > low);
        int first = stream.randInt(low, high - 1);
        int second = stream.randInt(low, high - 1);

        while (first == second) {
            second = stream.randInt(low, high - 1);
        }
        return new Pair(first, second);
    }


    /**
     * @param population
     * @param numToSelect
     * @param prng
     * @return
     */
    ArrayList<Solution> binary_tournament_selection(ArrayList<Solution> population,
                                                    int numToSelect,
                                                    RngStream prng) {
        // classical binary tournament selection
        assert (numToSelect > 0);
        Pair<Integer, Integer> indices;
        ArrayList<Solution> selected = new ArrayList<>();
        int pop_size = population.size() - 1;

        double score1, score2;
        for (int index = 0; index != numToSelect; ++index) {
            indices = select2(0, pop_size, prng);
            score1 = population.get(indices.getFirst()).getScore();
            score2 = population.get(indices.getSecond()).getScore();
            if (score1 < score2) {
                selected.add(new Solution(population.get(indices.getFirst())));
            } else {
                selected.add(new Solution(population.get(indices.getSecond())));
            }
        }
        return selected;
    }


    /**
     * solve the consensus problem
     * using an hybrid strategy (genetic algorithm + steepest descent)
     *
     * @param distanceMatrix
     * @param ranges
     * @return
     */
    public ArrayList<Solution> solve(double[][] distanceMatrix,
                                     ArrayList<Pair<Integer, Integer>> ranges) {
        assert (populationSize > 0);
        assert (numGenerations > 0);
        assert (eliteSize >= 0 && eliteSize < populationSize);
        assert (0. <= crossoverProbability && crossoverProbability <= 1.);
        assert (0. <= crossoverMixingRatio && crossoverMixingRatio <= 1.);
        assert (0. <= mutationProbability  && mutationProbability <= 1.);

        // seed the pseudorandom generator (MRG32k3a from L'Ecuyer)
        RngStream prng = new RngStream();
        prng.setSeed(seeds);

        // some declarations for later
        ArrayList<ArrayList<Solution>> best_solutions = new ArrayList<>();
        ArrayList<Solution> current_best_solutions;
        ArrayList<Solution> children, parents, elite;
        Solution parent1;
        Solution parent2;
        double current_best_score = Double.POSITIVE_INFINITY;
        double scaled_threshold = tolerance * ranges.size() * (ranges.size() - 1);


        // start the progress meter
        ProgressBar bar = new ProgressBar("", 40);

        // initialize the population
        ArrayList<Solution> population = initializeRandomSolutions(ranges, populationSize, prng);

        // main loop
        for (int generation_index = 0; generation_index != numGenerations; ++generation_index) {

            // output the progress bar if not silent
            if (verbose) {
                bar.update(((float) generation_index) / numGenerations);
            }

            // score the solutions and sort the population by score
            for (Solution solution : population) {
                assignPairwiseDistanceScore(solution, distanceMatrix);
            }
            Collections.sort(population);

            // remember the best solutions of the current generation
            current_best_solutions = new ArrayList<>();
            current_best_score = population.get(0).getScore(); // because it is sorted
            for (Solution solution : population) {
                if (solution.getScore() <= current_best_score + scaled_threshold) {
                    // if not elite, put it there
                    if (!current_best_solutions.contains(solution)) {
                        current_best_solutions.add(new Solution(solution));
                    }
                } else {
                    break;
                }
            }

            best_solutions.add(current_best_solutions);

            // elitist selection with only unique individuals, no repetition
            elite = new ArrayList<>();

            for (Solution solution : population) {
                if (elite.size() >= eliteSize) {
                    break;
                } else {
                    if (!elite.contains(solution)) {
                        elite.add(new Solution(solution));
                    }
                }
            }

            // selection process
            parents = binary_tournament_selection(population, ((populationSize - elite.size()) * 2), prng);
            children = new ArrayList<>();
            for (int i = 0; i != populationSize - eliteSize; ++i) {
                parent1 = parents.get(i * 2);
                parent2 = parents.get((i * 2) + 1);
                Solution child;

                // crossover
                if (prng.randU01() < crossoverProbability)
                {
                    child = uniformCrossover(parent1, parent2, crossoverMixingRatio, prng);
                }
                else
                {
                    child = new Solution(parent1);
                }

                // mutation
                if (prng.randU01() < mutationProbability)
                {
                    uniformMutate(child, ranges, mutationStrength, prng);
                }


                // improvement
                if (prng.randU01() < improvementProbability)
                    steepestDescent(child, distanceMatrix, ranges, improvementDepth);

                // children is complete
                children.add(child);
            }

            // replace the population by its children and the previous elite
            for (Solution solution : elite) {
                children.add(new Solution(solution));
            }

            // swap the two populations
            population = children;
        }

        // clean the progress bar
        if (verbose)
            bar.clean();


        // keep all the unique best solutions up to a specified suboptimal threshold
        ArrayList<Solution> suitable_solutions = new ArrayList<>();
        double scoreThreshold = current_best_score + scaled_threshold;
        for (ArrayList<Solution> bestSolutions : best_solutions) {
            for (Solution solution : bestSolutions) {
                if ((solution.getScore() < scoreThreshold) &&
                        (!suitable_solutions.contains(solution))){
                    suitable_solutions.add(new Solution(solution));
                }
            }
        }
        return suitable_solutions;
    }
}
